File: src/jsonrpc/eth/signTransaction/eth_signTransaction.js
/**
 * @fileoverview eth_signTransaction JSON-RPC method
 */

import { createRequest } from "../../types/JsonRpcRequest.js";

/**
 * @typedef {import('../../types/index.js').Address} Address
 * @typedef {import('../../types/index.js').Hash} Hash
 * @typedef {import('../../types/index.js').Quantity} Quantity
 * @typedef {import('../../types/index.js').BlockTag} BlockTag
 * @typedef {import('../../types/index.js').BlockSpec} BlockSpec
 * @typedef {import('../../types/JsonRpcRequest.js').JsonRpcRequest} JsonRpcRequest
 */

/**
 * Returns an RLP encoded transaction signed by the specified account.
 *
 * @example
 * Transaction: ...
 * Result: "0xa3f20717a250c2b0b729b7e5becbff67fdaef7e0699da4de7ca5895b02a170a12d887fd3b17bfdce3481f10bea41f45ba9f709d39ce8325427b57afcfc994cee1b"
 *
 * Implements the `eth_signTransaction` JSON-RPC method.
 */

/** The JSON-RPC method name */
export const method = "eth_signTransaction";
/**
 * Result for `eth_signTransaction`
 *
 * hex encoded bytes
 *
 * @typedef {Quantity} Result
 */

/**
 * Request for `eth_signTransaction`
 *
 * @typedef {JsonRpcRequest<'eth_signTransaction', [any]>} Request
 */

/**
 * Creates a eth_signTransaction JSON-RPC request
 *
 * @param {any} address
 * @param {number | string | null} [id] - Optional request ID
 * @returns {Request}
 */
export function SignTransactionRequest(address, id = null) {
	return /** @type {Request} */ (createRequest(method, [address], id));
}

File: src/jsonrpc/eth/signTransaction/eth_signTransaction.zig
const std = @import("std");
const types = @import("../../types.zig");

/// Returns an RLP encoded transaction signed by the specified account.
///
/// Example:
/// Transaction: ...
/// Result: "0xa3f20717a250c2b0b729b7e5becbff67fdaef7e0699da4de7ca5895b02a170a12d887fd3b17bfdce3481f10bea41f45ba9f709d39ce8325427b57afcfc994cee1b"
///
/// Implements the `eth_signTransaction` JSON-RPC method.
pub const EthSignTransaction = @This();

/// The JSON-RPC method name
pub const method = "eth_signTransaction";

/// Parameters for `eth_signTransaction`
pub const Params = struct {
    /// Transaction object generic to all types
    transaction: types.Quantity,

    pub fn jsonStringify(self: Params, jws: *std.json.Stringify) !void {
        try jws.beginArray();
        try jws.write(self.transaction);
        try jws.endArray();
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Params {
        if (source != .array) return error.UnexpectedToken;
        if (source.array.items.len != 1) return error.InvalidParamCount;

        return Params{
            .transaction = try std.json.innerParseFromValue(types.Quantity, allocator, source.array.items[0], options),
        };
    }
};

/// Result for `eth_signTransaction`
pub const Result = struct {
    /// hex encoded bytes
    value: types.Quantity,

    pub fn jsonStringify(self: Result, jws: *std.json.Stringify) !void {
        try jws.write(self.value);
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Result {
        return Result{
            .value = try std.json.innerParseFromValue(types.Quantity, allocator, source, options),
        };
    }
};
